src/components/cursor-controller.js

```diff
@@ -3,39 +3,113 @@ const TARGET_TYPE_INTERACTABLE = 2;
 const TARGET_TYPE_UI = 4;
 const TARGET_TYPE_INTERACTABLE_OR_UI = TARGET_TYPE_INTERACTABLE | TARGET_TYPE_UI;
 
+/**
+ * Manages targeting and physical cursor location. Has the following responsibilities:
+ *
+ * - Tracking which entities in the scene can be targeted by the cursor (`objects`).
+ * - Performing a raycast per-frame or on-demand to identify which entity is being currently targeted.
+ * - Updating the visual presentation and position of the `cursor` entity and `line` component per frame.
+ * - Sending an event when an entity is targeted or un-targeted.
+ */
 AFRAME.registerComponent("cursor-controller", {
-  dependencies: ["raycaster", "line"],
+  dependencies: ["line"],
   schema: {
     cursor: { type: "selector" },
     camera: { type: "selector" },
-    maxDistance: { default: 3 },
-    minDistance: { default: 0 },
+    far: { default: 3 },
+    near: { default: 0 },
     cursorColorHovered: { default: "#2F80ED" },
     cursorColorUnhovered: { default: "#FFFFFF" },
     rayObject: { type: "selector" },
-    useMousePos: { default: true },
-    drawLine: { default: false }
+    drawLine: { default: false },
+    objects: { default: "" }
   },
 
   init: function() {
     this.enabled = true;
-    this.inVR = false;
-    this.isMobile = AFRAME.utils.device.isMobile();
     this.currentTargetType = TARGET_TYPE_NONE;
-    this.currentDistance = this.data.maxDistance;
+    this.currentDistance = this.data.far;
     this.currentDistanceMod = 0;
     this.mousePos = new THREE.Vector2();
     this.wasCursorHovered = false;
-    this.origin = new THREE.Vector3();
-    this.direction = new THREE.Vector3();
-    this.raycasterAttr = this.el.getAttribute("raycaster");
-    this.controllerQuaternion = new THREE.Quaternion();
     this.data.cursor.setAttribute("material", { color: this.data.cursorColorUnhovered });
 
     this._handleCursorLoaded = this._handleCursorLoaded.bind(this);
     this.data.cursor.addEventListener("loaded", this._handleCursorLoaded);
+
+    // raycaster state
+    this.targets = [];
+    this.intersection = null;
+    this.raycaster = new THREE.Raycaster();
+    this.setDirty = this.setDirty.bind(this);
+    this.dirty = true;
+  },
+
+  update: function() {
+    this.raycaster.far = this.data.far;
+    this.raycaster.near = this.data.near;
+    this.setDirty();
   },
 
+  play: function() {
+    this.observer = new MutationObserver(this.setDirty);
```

Comment:
TIL `MutationObserver` exists.

Reply:
Me too. By the way, I don't know whether it's actually important for us to observe attribute changes or not -- that behavior is just inherited from the A-Frame raycaster. Do you know? That is, do we anticipate cases where an entity changes from raycast-targetable to not-raycast-targetable or vice versa?

Reply:
Right now we do. For example when you interact with the spawner, it creates a new duck and hides the spawner and prevents you from interacting with the spawner a second time on accident by changing its css attribute. See https://github.com/mozilla/hubs/blob/master/src/components/super-spawner.js#L188

Reply:
Cute! I didn't know exactly how spawners worked.